src/data/polynomial/ring_division.lean

```diff
@@ -987,6 +991,57 @@ begin
   { exact q }, { rw mul_comm, exact q },
 end
 
+lemma monic.irreducible_iff {R} [comm_semiring R] [no_zero_divisors R] {p : R[X]} (hp : p.monic) :
+  irreducible p ↔ p ≠ 1 ∧ ∀ a b, p = a * b →
+    a.monic → b.monic → 0 < a.nat_degree → 0 < b.nat_degree →
+    a.nat_degree + b.nat_degree = p.nat_degree → is_unit a ∨ is_unit b :=
```

Comment:
The idea is to make the RHS as weak as possible, because the weaker it is, the easier it is to verify, and an easy-to-verify criterion is what we want. As you noted, I could replace `is_unit a ∨ is_unit b` by `false`, but that makes the RHS stronger and the reverse implication harder to prove. Similarly, adding more assumptions/conditions like those about `nat_degree` also makes the RHS weaker, and the three `nat_degree` conditions makes it easy to show `a.nat_degree = b.nat_degree = 1` when `p.nat_degree = 2` in the quadratic case below.

Thanks for letting me know about [polynomial.irreducible_of_monic](https://leanprover-community.github.io/mathlib_docs/data/polynomial/field_division.html#polynomial.irreducible_of_monic), that could probably be golfed using this.

Reply:
I pushed a generalized of `polynomial.irreducible_of_monic` to #17696.

Reply:
I agree that restricting the RHS of the iff can make it easier to verify the condition, but we should also avoid irrelevant hypotheses that clutter up the state. As an extreme, we shouldn't include `0 ≠ 1` in the conditions even though we make use of it `not_irreducible_iff_exists_add_mul_eq_coeff`. Summarized, I believe we should avoid conclusions of the form `a ↔ b ∧ c` if it's easy to show that `b → c`.